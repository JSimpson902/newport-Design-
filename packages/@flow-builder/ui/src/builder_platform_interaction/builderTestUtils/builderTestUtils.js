export * from './screenEditorTestUtils.js';
export * from './commonTestUtils';
export * from './deepFindMatchers';
export * from './useAdvancedOptionsTestUtils';
export * from './goldObjectMatchers';
export * from './domTestUtils';
export * from './complexTypeTestUtils';
