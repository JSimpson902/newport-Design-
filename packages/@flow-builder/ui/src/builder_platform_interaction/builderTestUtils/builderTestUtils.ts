export * from './combobox';
export * from './commonTestUtils';
export * from './complexTypeTestUtils';
export * from './deepFindMatchers';
export * from './domTestUtils';
export * from './events';
export * from './goldObjectMatchers';
export * from './manuallyAssignVariablesCheckboxTestUtils';
export * from './screenEditorTestUtils';
export * from './selectors';

// @sa11y testing appears to take some extra time (beyond the given 5000ms default jest timeout)
export const ACCESSIBILITY_TEST_TIMEOUT_IN_MS = 15000;

/**
 * Mocks a label in jest
 *
 * @param jest - The jest instance
 * @param labelName - The label name
 * @returns the mock label
 */
export function mockLabel(jest, labelName: string) {
    jest.mock(
        `@salesforce/label/${labelName}`,
        () => {
            return { default: `${labelName} mock value` };
        },
        { virtual: true }
    );

    return `${labelName} mock value`;
}
