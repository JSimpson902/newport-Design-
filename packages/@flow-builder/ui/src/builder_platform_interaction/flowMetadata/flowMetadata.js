export const FLOW_PROCESS_TYPE = {
    AUTO_LAUNCHED_FLOW : 'AutoLaunchedFlow',
    FLOW : 'Flow'
};

export const ELEMENT_TYPE = {
    ACTION_CALL: 'ACTION_CALL',
    ASSIGNMENT: 'ASSIGNMENT',
    VARIABLE: 'VARIABLE',
    DECISION: 'DECISION',
    DECISION_WITH_MODIFIED_AND_DELETED_OUTCOMES: 'DECISION_WITH_MODIFIED_AND_DELETED_OUTCOMES',
    APEX_PLUGIN_CALL: 'APEX_PLUGIN_CALL',
    APEX_CALL: 'APEX_CALL',
    EMAIL_ALERT: 'EMAIL_ALERT',
    CHOICE: 'CHOICE',
    CONSTANT: 'CONSTANT',
    DYNAMIC_CHOICE_SET: 'DYNAMIC_CHOICE_SET',
    FORMULA: 'FORMULA',
    STAGE: 'STAGE',
    TEXT_TEMPLATE: 'TEXT_TEMPLATE',
    LOOP: 'LOOP',
    OUTCOME: 'OUTCOME',
    RECORD_CREATE: 'RECORD_CREATE',
    RECORD_DELETE: 'RECORD_DELETE',
    RECORD_LOOKUP: 'RECORD_LOOKUP',
    RECORD_UPDATE: 'RECORD_UPDATE',
    SCREEN: 'SCREEN',
    STEP: 'STEP',
    SUBFLOW: 'SUBFLOW',
    START_ELEMENT: 'START_ELEMENT',
    WAIT: 'WAIT',
    FLOW_PROPERTIES: 'FLOW_PROPERTIES',
    DEFAULT: 'defaultElement'
};

export const CONNECTOR_TYPE = {
    REGULAR: 'REGULAR',
    FAULT: 'FAULT',
    DEFAULT: 'DEFAULT',
    START: 'START',
    LOOP_NEXT: 'LOOP_NEXT',
    LOOP_END: 'LOOP_END'
};

export const SUB_ELEMENT_TYPE = {
    ASSIGNMENT_ITEM: 'ASSIGNMENT_ITEM',
    CONDITION: 'CONDITION',
    RECORD_LOOKUP_FILTER_ITEM: 'RECORD_LOOKUP_FILTER_ITEM',
    RECORD_LOOKUP_FIELD: 'RECORD_LOOKUP_FIELD',
};

export const ACTION_TYPE = {
    APEX: 'apex',
    EMAIL_ALERT: 'emailAlert',
    QUICK_ACTION: 'quickAction',
    FLOW: 'flow',
    COMPONENT: 'component'
};

export const METADATA_KEY = {
    ACTION_CALLS: 'actionCalls',
    APEX_PLUGIN_CALLS: 'apexPluginCalls',
    ASSIGNMENTS: 'assignments',
    DECISIONS: 'decisions',
    WAITS: 'waits',
    FORMULAS: 'formulas',
    VARIABLES: 'variables',
    CONSTANTS: 'constants',
    SUBFLOWS: 'subflows',
    RECORD_CREATE: 'recordCreates',
    RECORD_LOOKUP: 'recordLookups',
    RECORD_DELETE: 'recordDeletes',
    RECORD_UPDATE: 'recordUpdates',
    LOOPS: 'loops',
    SCREENS: 'screens'
};

/**
 * AND and OR are flow values which are used by the backend as well as being as used
 * on the frontend (combobox values).  CUSTOM_LOGIC is used only by the frontend as a
 * combobox value
 * @type {{AND: string, OR: string, ALWAYS: string, CUSTOM_LOGIC: string}}
 */
export const CONDITION_LOGIC = {
    AND: 'and',
    OR: 'or',
    CUSTOM_LOGIC: 'custom_logic',
    NO_CONDITIONS: 'no_conditions'
};

/**
 * The top level process metadata values to save with each flow. This includes the BuilderType
 * which helps us identify flow versions created by the Lightning Flow Builder.
 */
export const PROCESS_METADATA_VALUES = [
    {
        name: 'BuilderType',
        value: {
            stringValue: 'LightningFlowBuilder'
        }
    }
];

/**
 * List of Flow metadata fields that can be references to other Flow elements
 * i.e. fields whose values need to be converted from dev names to GUIDs when loading the flow into the UI client and vice versa.
 */
export const REFERENCE_FIELDS = new Set([
    'elementReference',
    'targetReference',
    'assignToReference',
    'assignNextValueToReference',
    'collectionReference',
    'inputReference',
    'outputReference',
    'assignRecordIdToReference',
    'leftValueReference',
    'outcomeReference',
    'childReference',
    'leftHandSide',
    'defaultValue',
    'source',
    'target',
    'rightHandSideGuid'
]);

/**
 * List of Flow metadata fields that can contain {!reference} tags referencing flow elements
 * These would also need to be replaced
 */
export const TEMPLATE_FIELDS = new Set([
    // TODO fill out more values and confirm that all FEROV stringValue fields allow this behaviour
    'stringValue', // field of a ferov
    'rightHandSide', // field of a ferov
    'expression',  // represents body of a formula
    'text',        // represents body of a text template
    'fieldText',   // body of screen field of type Display Text
    'helpText',     // help text for a screen or screen field
    'interviewLabel' // interview label for the flow properties
]);

export const EXPRESSION_RE = /\{!([^}]+)\}/g;
