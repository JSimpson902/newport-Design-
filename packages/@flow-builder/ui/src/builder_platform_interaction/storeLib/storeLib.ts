// @ts-nocheck
export { combinedReducer } from './combinedReducer';
export { createSelector } from './createSelector';
export { Store } from './store';
export { generateGuid } from './guidGenerator';
export { deepCopy } from './deepCopy';
export { isPlainObject } from './isPlainObject';
