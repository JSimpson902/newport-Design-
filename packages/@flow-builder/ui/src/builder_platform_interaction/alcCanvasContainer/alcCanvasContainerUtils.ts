import {
    getConfigForElementType,
    getChildElementTypesWithOverridenProperties
} from 'builder_platform_interaction/elementConfig';
import { ELEMENT_TYPE, FLOW_TRIGGER_TYPE } from 'builder_platform_interaction/flowMetadata';
import {
    getAlcElementType,
    startElementDescription,
    hasTrigger,
    hasContext,
    isRecordTriggeredFlow
} from 'builder_platform_interaction/alcCanvasUtils';

import { ElementMetadata } from 'builder_platform_interaction/autoLayoutCanvas';

/**
 * Checks if an element can have a fault connector
 *
 * @param startElement - the start element
 * @param metadata - the element metadata
 * @returns true if the element can hava a fault connector, false otherwise
 */
function canHaveFaultConnector(startElement: UI.Start, metadata: UI.ElementConfig) {
    // W-8985288: clean this up to make canHaveFaultConnector more dynamic than a static attribute in elementConfig
    if (
        metadata.elementType === ELEMENT_TYPE.RECORD_UPDATE &&
        startElement.triggerType === FLOW_TRIGGER_TYPE.BEFORE_SAVE
    ) {
        return false;
    }

    return metadata.canHaveFaultConnector;
}

/**
 * Augments the elements metadata with auto layout specific elements and properties
 *
 * @param elementsMetadata - The elements metadata
 * @param startElement - The start element
 * @returns the augmented elements metadata for the auto layout canvas
 */
export function augmentElementsMetadata(elementsMetadata: UI.ElementConfig[], startElement: UI.Start) {
    const { ROOT_ELEMENT, START_ELEMENT, END_ELEMENT } = ELEMENT_TYPE;

    // @ts-ignore
    const alcElementsMetadata: ElementMetadata[] = elementsMetadata.map((metadata) => ({
        ...metadata,
        canHaveFaultConnector: canHaveFaultConnector(startElement, metadata),
        type: getAlcElementType(metadata.elementType)
    }));

    getChildElementTypesWithOverridenProperties().forEach((elementType) => {
        const elementConfig = getConfigForElementType(elementType);
        const nodeConfig = elementConfig.nodeConfig!;

        const { iconName, iconBackgroundColor } = nodeConfig;
        const labels = elementConfig.labels!;

        alcElementsMetadata.push({
            icon: iconName,
            iconBackgroundColor: iconBackgroundColor!,
            label: labels.singular,
            elementType,
            type: getAlcElementType(elementType),
            // @ts-ignore  TODO: do we need this in the metadata?
            canHaveFaultConnector: elementConfig.canHaveFaultConnector
        });
    });

    const endElement: UI.ElementConfig = getConfigForElementType(END_ELEMENT);
    let nodeConfig = endElement.nodeConfig!;

    const endElementConfig = {
        section: nodeConfig.section,
        icon: nodeConfig.iconName,
        iconBackgroundColor: nodeConfig.iconBackgroundColor,
        iconShape: nodeConfig.iconShape,
        iconSize: nodeConfig.iconSize,
        description: nodeConfig.description,
        label: endElement.labels!.singular,
        value: END_ELEMENT,
        elementType: END_ELEMENT,
        type: getAlcElementType(END_ELEMENT),
        canHaveFaultConnector: false
    };

    let startElementConfig: any = getConfigForElementType(START_ELEMENT);
    nodeConfig = startElementConfig.nodeConfig!;

    startElementConfig = {
        description: startElementDescription(startElement.triggerType!),
        icon: nodeConfig.iconName,
        iconBackgroundColor: nodeConfig.iconBackgroundColor,
        iconShape: nodeConfig.iconShape,
        iconSize: nodeConfig.iconSize,
        label: startElementConfig.labels!.singular,
        value: START_ELEMENT,
        elementType: START_ELEMENT,
        type: getAlcElementType(START_ELEMENT),
        canHaveFaultConnector: false,
        hasTrigger: hasTrigger(startElement.triggerType),
        hasContext: hasContext(startElement.triggerType),
        isRecordTriggeredFlow: isRecordTriggeredFlow(startElement.triggerType),
        menuComponent: 'builder_platform_interaction/alcStartMenu'
    };

    return [
        ...alcElementsMetadata,
        {
            section: null,
            icon: '',
            label: '',
            elementType: ROOT_ELEMENT,
            value: ROOT_ELEMENT,
            type: getAlcElementType(ROOT_ELEMENT),
            canHaveFaultConnector: false
        },
        startElementConfig,
        endElementConfig
    ];
}
