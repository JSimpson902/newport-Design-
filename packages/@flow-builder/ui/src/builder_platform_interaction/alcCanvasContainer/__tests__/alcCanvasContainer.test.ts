// @ts-nocheck
import { createComponent, ticks } from 'builder_platform_interaction/builderTestUtils';
import { getChildElementTypesWithOverridenProperties } from 'builder_platform_interaction/elementConfig';
import { ELEMENT_TYPE } from 'builder_platform_interaction/flowMetadata';
import { Store } from 'builder_platform_interaction/storeLib';
import { augmentElementsMetadata } from '../alcCanvasContainerUtils';

jest.mock('builder_platform_interaction/alcCanvas', () => require('builder_platform_interaction_mocks/alcCanvas'));

jest.mock('builder_platform_interaction/storeLib');

const elementsMetadata = [
    {
        section: 'Interaction',
        type: 'default',
        icon: 'standard:screen',
        label: 'Screen',
        value: 'Screen',
        elementType: 'Screen',
        description: 'Collect information from users who run the flow or show them some information.',
        canHaveFaultConnector: false
    },
    {
        section: 'Interaction',
        type: 'default',
        icon: 'standard:lightning_component',
        label: 'New Action',
        value: 'ActionCall',
        elementType: 'ActionCall',
        description:
            'Perform an action outside of the flow. Choose from your orgâ€™s custom create and update actions or an out-of-the-box action, like Post to Chatter or Submit for Approval.',
        canHaveFaultConnector: true
    },
    {
        section: 'Flow Control',
        type: 'branch',
        icon: 'standard:decision',
        label: 'Decision',
        value: 'Decision',
        elementType: 'Decision',
        description: 'Create paths for the flow to take based on conditions you set.',
        canHaveFaultConnector: false
    },
    {
        section: 'Flow Control',
        type: 'branch',
        icon: 'standard:waits',
        label: 'Pause',
        value: 'Wait',
        elementType: 'Wait',
        description:
            'Pause the flow. Resume the flow when the org receives a platform event message or a specified day, date, or time occurs.',
        canHaveFaultConnector: true
    },
    {
        section: 'Data Operation',
        type: 'default',
        icon: 'standard:record_create',
        label: 'RecordCreate',
        value: 'RecordCreate',
        elementType: 'RecordCreate',
        description: 'Create Salesforce records using values from the flow.',
        canHaveFaultConnector: true
    },
    {
        section: 'Data Operation',
        type: 'default',
        icon: 'standard:record_update',
        label: 'RecordUpdate',
        value: 'RecordUpdate',
        elementType: 'RecordUpdate',
        description: 'Find Salesforce records, and store their field values to use later in the flow.',
        canHaveFaultConnector: true
    },
    {
        section: 'Data Operation',
        type: 'default',
        icon: 'standard:record_lookup',
        label: 'RecordQuery',
        value: 'RecordQuery',
        elementType: 'RecordQuery',
        description: 'Update Salesforce records using values from the flow.',
        canHaveFaultConnector: true
    },
    {
        section: 'Data Operation',
        type: 'default',
        icon: 'standard:record_delete',
        label: 'RecordDelete',
        value: 'RecordDelete',
        elementType: 'RecordDelete',
        description: 'Delete Salesforce records.',
        canHaveFaultConnector: true
    },
    {
        section: null,
        icon: 'utility:right',
        label: 'Start',
        value: 'START_ELEMENT',
        elementType: 'START_ELEMENT',
        type: 'start',
        canHaveFaultConnector: false
    },
    {
        section: null,
        icon: '',
        label: '',
        elementType: 'root',
        value: 'root',
        type: 'root',
        canHaveFaultConnector: false
    },
    {
        section: 'Logic',
        icon: 'standard:first_non_empty',
        description: 'End Description',
        label: 'End',
        value: 'END_ELEMENT',
        elementType: 'END_ELEMENT',
        type: 'end',
        canHaveFaultConnector: false
    },
    {
        section: 'Interaction',
        type: 'orchestratedstage',
        icon: 'standard:screen',
        label: 'orchestratedStage',
        value: 'orchestratedstage',
        elementType: 'orchestratedstage',
        description: 'stepped stage desc',
        canHaveFaultConnector: false,

        dynamicNodeComponent: 'builder_platform_interaction/orchestratedStageNode'
    }
];

const startElement = {
    elementType: 'START_ELEMENT'
};
const createComponentForTest = async (optionsOverrides = {}) => {
    return createComponent(
        'builder_platform_interaction-alc-canvas-container',
        {
            elementsMetadata,
            isSelectionMode: false,
            numPasteElementsAvailable: 0
        },
        optionsOverrides
    );
};

describe('alc canvas container', () => {
    let cmp;

    beforeAll(() => {
        Store.getStore.mockImplementation(() => {
            return {
                getCurrentState: jest.fn(() => {
                    return {
                        elements: {
                            startGuid: startElement,
                            root: { elementType: ELEMENT_TYPE.ROOT_ELEMENT }
                        },
                        properties: { processType: 'autolaunched', isAutoLayoutCanvas: true }
                    };
                }),
                subscribe: jest.fn(() => jest.fn()),
                unsubscribe: jest.fn()
            };
        });
    });

    beforeEach(async () => {
        cmp = await createComponentForTest();
    });

    const getBuilderContainerElement = () =>
        cmp.shadowRoot.querySelector('builder_platform_interaction-alc-canvas-container');

    const getAlcCanvas = () => cmp.shadowRoot.querySelector('builder_platform_interaction-alc-canvas');

    it('renders the component', () => {
        expect(getBuilderContainerElement).not.toBeNull();
    });

    it('offsets when in selection mode', async () => {
        expect(getAlcCanvas().offsets).toEqual([0, 58]);

        cmp.isSelectionMode = true;
        await ticks(1);
        expect(getAlcCanvas().offsets).toEqual([320, 58]);
    });

    it('intializes augmented metadata types correctly', () => {
        const elementTypes = cmp.elementsMetadata.map((element) => {
            return element.elementType;
        });
        const expectedElementTypeArray = [
            ELEMENT_TYPE.END_ELEMENT,
            ELEMENT_TYPE.ROOT_ELEMENT,
            ELEMENT_TYPE.START_ELEMENT
        ].concat(getChildElementTypesWithOverridenProperties());

        expect(elementTypes).toEqual(expect.arrayContaining(expectedElementTypeArray));
    });

    it('can have a fault connector for record update', () => {
        const elementsMetadata = cmp.elementsMetadata.filter(
            (element) => element.elementType === ELEMENT_TYPE.RECORD_UPDATE
        );
        expect(elementsMetadata.length).toBe(1);
        expect(elementsMetadata[0].canHaveFaultConnector).toBe(true);
    });

    it('generates the correct connector menu metadata ', () => {
        const expectedElementTypes = new Set([
            'Screen',
            'ActionCall',
            'Decision',
            'Wait',
            'RecordCreate',
            'RecordUpdate',
            'RecordQuery',
            'RecordDelete',
            'START_ELEMENT',
            'root',
            'END_ELEMENT',
            'orchestratedstage',
            'EMAIL_ALERT',
            'APEX_CALL',
            'ApexPlugin',
            'EXTERNAL_SERVICE'
        ]);

        const { menuComponent, elementTypes } = getAlcCanvas().connectorMenuMetadata;

        expect(menuComponent).toEqual('builder_platform_interaction/alcConnectorMenu');
        expect(elementTypes).toEqual(expectedElementTypes);
    });

    it('augments the metadata', () => {
        expect(augmentElementsMetadata(elementsMetadata, startElement)).toMatchSnapshot();
    });
});

describe('before save flow', () => {
    let cmp;

    beforeAll(() => {
        Store.getStore.mockImplementation(() => {
            return {
                getCurrentState: jest.fn(() => {
                    return {
                        elements: { startGuid: { elementType: 'START_ELEMENT', triggerType: 'RecordBeforeSave' } },
                        properties: { processType: 'autolaunched' }
                    };
                }),
                subscribe: jest.fn(() => jest.fn()),
                unsubscribe: jest.fn()
            };
        });
    });

    beforeEach(async () => {
        cmp = await createComponentForTest();
    });

    it('cannot have a fault connector for record update', async () => {
        const elementsMetadata = cmp.elementsMetadata.filter(
            (element) => element.elementType === ELEMENT_TYPE.RECORD_UPDATE
        );
        expect(elementsMetadata.length).toBe(1);
        expect(elementsMetadata[0].canHaveFaultConnector).toBe(false);
    });
});
