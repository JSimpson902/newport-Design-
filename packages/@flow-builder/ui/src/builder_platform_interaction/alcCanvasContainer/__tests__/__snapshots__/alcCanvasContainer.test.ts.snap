// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`alc canvas container augments the metadata 1`] = `
Array [
  Object {
    "canHaveFaultConnector": false,
    "description": "Collect information from users who run the flow or show them some information.",
    "elementType": "Screen",
    "icon": "standard:screen",
    "label": "Screen",
    "section": "Interaction",
    "type": "default",
    "value": "Screen",
  },
  Object {
    "canHaveFaultConnector": true,
    "description": "Perform an action outside of the flow. Choose from your orgâ€™s custom create and update actions or an out-of-the-box action, like Post to Chatter or Submit for Approval.",
    "elementType": "ActionCall",
    "icon": "standard:lightning_component",
    "label": "New Action",
    "section": "Interaction",
    "type": "default",
    "value": "ActionCall",
  },
  Object {
    "canHaveFaultConnector": false,
    "description": "Create paths for the flow to take based on conditions you set.",
    "elementType": "Decision",
    "icon": "standard:decision",
    "label": "Decision",
    "section": "Flow Control",
    "type": "branch",
    "value": "Decision",
  },
  Object {
    "canHaveFaultConnector": true,
    "description": "Pause the flow. Resume the flow when the org receives a platform event message or a specified day, date, or time occurs.",
    "elementType": "Wait",
    "icon": "standard:waits",
    "label": "Pause",
    "section": "Flow Control",
    "type": "branch",
    "value": "Wait",
  },
  Object {
    "canHaveFaultConnector": true,
    "description": "Create Salesforce records using values from the flow.",
    "elementType": "RecordCreate",
    "icon": "standard:record_create",
    "label": "RecordCreate",
    "section": "Data Operation",
    "type": "default",
    "value": "RecordCreate",
  },
  Object {
    "canHaveFaultConnector": true,
    "description": "Find Salesforce records, and store their field values to use later in the flow.",
    "elementType": "RecordUpdate",
    "icon": "standard:record_update",
    "label": "RecordUpdate",
    "section": "Data Operation",
    "type": "default",
    "value": "RecordUpdate",
  },
  Object {
    "canHaveFaultConnector": true,
    "description": "Update Salesforce records using values from the flow.",
    "elementType": "RecordQuery",
    "icon": "standard:record_lookup",
    "label": "RecordQuery",
    "section": "Data Operation",
    "type": "default",
    "value": "RecordQuery",
  },
  Object {
    "canHaveFaultConnector": true,
    "description": "Delete Salesforce records.",
    "elementType": "RecordDelete",
    "icon": "standard:record_delete",
    "label": "RecordDelete",
    "section": "Data Operation",
    "type": "default",
    "value": "RecordDelete",
  },
  Object {
    "canHaveFaultConnector": false,
    "elementType": "START_ELEMENT",
    "icon": "utility:right",
    "label": "Start",
    "section": null,
    "type": "start",
    "value": "START_ELEMENT",
  },
  Object {
    "canHaveFaultConnector": false,
    "elementType": "root",
    "icon": "",
    "label": "",
    "section": null,
    "type": "root",
    "value": "root",
  },
  Object {
    "canHaveFaultConnector": false,
    "description": "End Description",
    "elementType": "END_ELEMENT",
    "icon": "standard:first_non_empty",
    "label": "End",
    "section": "Logic",
    "type": "end",
    "value": "END_ELEMENT",
  },
  Object {
    "canHaveFaultConnector": false,
    "description": "stepped stage desc",
    "dynamicNodeComponent": "builder_platform_interaction/orchestratedStageNode",
    "elementType": "orchestratedstage",
    "icon": "standard:screen",
    "label": "orchestratedStage",
    "section": "Interaction",
    "type": "default",
    "value": "orchestratedstage",
  },
  Object {
    "canHaveFaultConnector": undefined,
    "elementType": "EMAIL_ALERT",
    "icon": "standard:email",
    "iconBackgroundColor": "background-navy",
    "label": "FlowBuilderElementConfig.emailAlertSingularLabel",
    "menuComponent": "builder_platform_interaction/alcNodeMenu",
    "type": "default",
  },
  Object {
    "canHaveFaultConnector": undefined,
    "elementType": "APEX_CALL",
    "icon": "standard:apex",
    "iconBackgroundColor": "background-navy",
    "label": "FlowBuilderElementConfig.apexSingularLabel",
    "menuComponent": "builder_platform_interaction/alcNodeMenu",
    "type": "default",
  },
  Object {
    "canHaveFaultConnector": undefined,
    "elementType": "ApexPlugin",
    "icon": "standard:apex_plugin",
    "iconBackgroundColor": "background-navy",
    "label": "FlowBuilderElementConfig.apexPluginSingularLabel",
    "menuComponent": "builder_platform_interaction/alcNodeMenu",
    "type": "default",
  },
  Object {
    "canHaveFaultConnector": undefined,
    "elementType": "EXTERNAL_SERVICE",
    "icon": "standard:custom_notification",
    "iconBackgroundColor": "background-navy",
    "label": "FlowBuilderElementConfig.actionSingularLabel",
    "menuComponent": "builder_platform_interaction/alcNodeMenu",
    "type": "default",
  },
  Object {
    "canHaveFaultConnector": false,
    "elementType": "root",
    "icon": "",
    "label": "",
    "section": null,
    "type": "root",
    "value": "root",
  },
  Object {
    "canHaveFaultConnector": false,
    "description": undefined,
    "elementType": "START_ELEMENT",
    "hasContext": true,
    "hasTrigger": true,
    "icon": "utility:right",
    "iconBackgroundColor": "background-green",
    "iconShape": "circle",
    "iconSize": "medium",
    "isRecordTriggeredFlow": false,
    "label": "FlowBuilderElementConfig.startElementSingularLabel",
    "menuComponent": "builder_platform_interaction/alcStartMenu",
    "type": "start",
    "value": "START_ELEMENT",
  },
  Object {
    "canHaveFaultConnector": false,
    "description": "FlowBuilderLeftPanelElements.endElementDescription",
    "elementType": "END_ELEMENT",
    "icon": "utility:stop",
    "iconBackgroundColor": "background-red",
    "iconShape": "circle",
    "iconSize": "medium",
    "label": "FlowBuilderElementConfig.endElementSingularLabel",
    "section": "FlowBuilderLeftPanelElements.flowControlLogicLabel",
    "type": "end",
    "value": "END_ELEMENT",
  },
]
`;
